crates/fuel-core/src/schema/tx/assemble_tx.rs

```diff
@@ -0,0 +1,879 @@
+            return Ok(self)
+        }
+
+        let memory = self.arguments.shared_memory_pool.get_memory().await;
+        let chain_id = self.arguments.consensus_parameters.chain_id();
+        self.tx
+            .precompute(&chain_id)
+            .map_err(|err| anyhow::anyhow!("{:?}", err))?;
+
+        let parameters =
+            CheckPredicateParams::from(self.arguments.consensus_parameters.as_ref());
+        let read_view = self.arguments.read_view.clone();
+
+        let mut tx_to_estimate = self.tx;
+        let estimated_tx = tokio_rayon::spawn_fifo(move || {
+            let result = tx_to_estimate.estimate_predicates(
+                &parameters,
+                memory,
+                read_view.as_ref(),
+            );
+            result.map(|_| tx_to_estimate)
+        })
+        .await
+        .map_err(|err| anyhow::anyhow!("{:?}", err))?;
+
+        self.tx = estimated_tx;
+
+        Ok(self)
+    }
+
+    async fn estimate_script_if_possible(&mut self) -> anyhow::Result<()> {
+        if !self.is_runnable_script() {
+            return Ok(())
+        }
+
+        let Some(script_ref) = self.tx.as_script_mut() else {
+            unreachable!("The transaction is a script, checked above; qed");
+        };
+
+        // Trick to avoid cloning `Script`
+        let dummy_script = Transaction::script(
+            Default::default(),
+            Default::default(),
+            Default::default(),
+            Default::default(),
+            Default::default(),
+            Default::default(),
+            Default::default(),
+        );
+        let mut script = core::mem::replace(script_ref, dummy_script);
+
+        *script.script_gas_limit_mut() = 0;
+
+        let dry_run_gas_price = 0;
+        set_max_fee(
+            &mut script,
+            &self.arguments.consensus_parameters,
+            self.arguments.reserve_gas,
+            dry_run_gas_price,
+            self.original_max_fee,
+        );
+
+        let max_tx_gas = self
+            .arguments
+            .consensus_parameters
+            .tx_params()
+            .max_gas_per_tx();
+
+        let has_spendable_input = script.inputs().iter().any(|input| match input {
+            Input::CoinSigned(_)
+            | Input::CoinPredicate(_)
+            | Input::MessageCoinSigned(_)
+            | Input::MessageCoinPredicate(_) => true,
+            Input::MessageDataSigned(_)
+            | Input::MessageDataPredicate(_)
+            | Input::Contract(_) => false,
+        });
+
+        // If `max_fee` policy is not set, set it because it affects the maximum gas
+        // used by transaction.
+        if script.policies().get(PolicyType::MaxFee).is_none() {
+            script.set_max_fee_limit(0);
+        }
+
+        let gas_costs = self.arguments.consensus_parameters.gas_costs();
+        let fee_params = self.arguments.consensus_parameters.fee_params();
+
+        let mut status;
+        let mut gas_used_by_tx;
+
+        loop {
```

Comment:
Can we document what's the goal of this loop ? it's unclear to me at first look why we can't dry_run once.

Reply:
We also have a dangling todo to use the continue on error feature, so I'll use that as well.